File: db-gate/Entrelares.DbContracts/Models/AppNotification.cs
using Supabase.Postgrest.Attributes;
using Supabase.Postgrest.Models;

namespace Entrelares.Models
{
    [Table("notifications")]
    public class AppNotification : BaseModel
    {
        [PrimaryKey("id", false)]
        public long Id { get; set; }

        [Column("recipient_profile_id")]
        public long RecipientProfileId { get; set; }

        [Column("type")]
        public string Type { get; set; } = string.Empty;

        [Column("title")]
        public string Title { get; set; } = string.Empty;

        [Column("message")]
        public string Message { get; set; } = string.Empty;

        // U-13: render data for NotificationRenderer — the row's stored
        // Title/Message are a PT-BR sentence written by whoever ACTED, while the
        // reader may speak another language. NULL on every row written before
        // the item; the renderer falls back to the stored text for those.
        //
        // TYPED AS JObject, NOT string (found by NotificationParamsTests): the
        // column is JSONB, so PostgREST returns a JSON OBJECT. Binding it to a
        // string makes Newtonsoft — the serializer the Supabase SDK uses — throw
        // `Unexpected character encountered while parsing value: {` on the WHOLE
        // response. That would not have failed the notification's own rendering:
        // it would have broken every client read of the table the moment the
        // first row carried params.
        [Column("params")]
        public Newtonsoft.Json.Linq.JObject? Params { get; set; }

        /// <summary>The payload as JSON text, for <c>NotificationRenderer</c>
        /// (which parses with System.Text.Json and treats null/invalid as
        /// "fall back to the stored sentence").
        ///
        /// <para><b>[JsonIgnore] is load-bearing.</b> The Postgrest model binder
        /// serializes every PUBLIC property, mapped or not — without this the
        /// column-less <c>ParamsJson</c> rode along on INSERTs and PostgREST
        /// answered <c>PGRST204: Could not find the 'ParamsJson' column</c>,
        /// breaking EVERY notification insert (including SwapRequestService's).
        /// "Not mapped" is not the default here; it has to be declared.</para></summary>
        [Newtonsoft.Json.JsonIgnore]
        public string? ParamsJson => Params?.ToString(Newtonsoft.Json.Formatting.None);

        [Column("swap_request_id")]
        public long? SwapRequestId { get; set; }

        [Column("is_read")]
        public bool IsRead { get; set; }

        [Column("created_at")]
        public DateTime CreatedAt { get; set; }
    }
}

File: packages/entrelares_db_contracts/lib/models/app_notification.dart
import 'dart:convert';

/// The `notifications` row. Mirrors `Entrelares/Models/AppNotification.cs`.
/// `params` is the U-13 render payload (VALUES only, never a sentence) that
/// [NotificationRenderer] rebuilds in the READER's language; the stored
/// `title`/`message` are the PT-BR fallback sentences.
class AppNotification {
  final int id;
  final int recipientProfileId;
  final String type;
  final String title;
  final String message;
  final Map<String, dynamic>? params;
  final int? swapRequestId;
  final bool isRead;
  final String? createdAt;

  const AppNotification({
    required this.id,
    required this.recipientProfileId,
    required this.type,
    required this.title,
    required this.message,
    this.params,
    this.swapRequestId,
    this.isRead = false,
    this.createdAt,
  });

  /// What `NotificationRenderer` takes — the raw JSON text, null when the
  /// row predates U-13 (legacy rows fall back to the stored sentence).
  String? get paramsJson => params == null ? null : jsonEncode(params);

  factory AppNotification.fromJson(Map<String, dynamic> json) =>
      AppNotification(
        id: json['id'] as int,
        recipientProfileId: json['recipient_profile_id'] as int,
        type: (json['type'] as String?) ?? '',
        title: (json['title'] as String?) ?? '',
        message: (json['message'] as String?) ?? '',
        params: (json['params'] as Map?)?.cast<String, dynamic>(),
        swapRequestId: json['swap_request_id'] as int?,
        isRead: (json['is_read'] as bool?) ?? false,
        createdAt: json['created_at'] as String?,
      );
}
